### PR TITLE
Add diff support

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
The last large refactoring I've planned for the `openssl_*`/`x509_*` modules: move the code from the _info modules to module_utils, so that the non-_info modules can use that code to generate a nice diff result.

(That's the current state.)

Will implement #38 for openssl_csr, openssl_privatekey, x509_certificate, x509_crl.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
openssl_csr
openssl_csr_info
openssl_privatekey
openssl_privatekey_info
x509_certificate
x509_certificate_info
x509_crl
x509_crl_info

### CODE REVIEW

**Comment by felixfontein** [comment]
For openssl_privatekey:
```.diff
TASK [(pyopenssl) Regenerate - adjust key type] **************************************************************************************************************
ok: [localhost] => (item=never)
failed: [localhost] (item=fail) => {"ansible_loop_var": "item", "changed": false, "item": "fail", "msg": "Key has wrong type and/or size. Will not proceed. To force regeneration, call the module with `generate` set to `partial_idempotence`, `full_idempotence` or `always`, or with `force=yes`."}
--- before
+++ after
@@ -2,26 +2,28 @@
     "can_parse_key": true,
     "key_is_consistent": true,
     "public_data": {
-        "exponent": 65537,
-        "modulus": 2459952894561635439598978347943110412418995662985613287404144212998554321326936560023559651032149885340714006860807559601233641445722717022079148118904488265693586503645226249086623201373927795112611634933577652989405482663249871162787624331925273493381094152484845087645394066747620625538238184628014306166579862831,
-        "size": 1048
+        "g": 124198426600828562094459997614787306694768831551010732972402687606330715505612989100180339716236730925676274746268944878371240071719613017675044530606625275483675978293907001274344722251970824353760403216729828522048875673912087170058977500186388937723403629770732335374977378825760854348197127653687580553293,
+        "p": 126448264294451924943718235079920701502781600385000069644750274760649246310848956715097417013794438451386931665063901502971993284176007629053601837189616348621415429228646313180819757235441694307847188847064790886732909309040163224176507151642978083998892638063163931764868453421654774926625588966944458795779,
+        "q": 754300365509485759907184804403109071699166915771,
+        "size": 1024,
+        "y": 29854936438211285703242148937706190096412434196347585744006585587091782645211184796513966286951279665700803084945006638541130711864989877437045321566808745942438772698462030360228048650690020677208922481485791112868234779250785791457558541977045127983951905023623666104226068630204521465134221794719506857638
     },
-    "public_key": "-----BEGIN PUBLIC KEY-----\nMIGiMA0GCSqGSIb3DQEBAQUAA4GQADCBjAKBhADQzOYydn3qTLCFsGLcRbMkdea1\nymA6I9T7VXsyzL3FAk1RjNHY/wHYX/mBGwxLeJOH1xRloGN2cGxlXzTJepg4oeWy\nzHIEmG64VEp2D8tXRgHzLspvjXHk2/BPUnra/+fSh4VsI7/3DZDem08CZlwcVL0U\niNROHjkyVWlC170KG2cdLwIDAQAB\n-----END PUBLIC KEY-----\n",
+    "public_key": "-----BEGIN PUBLIC KEY-----\nMIIBtzCCASwGByqGSM44BAEwggEfAoGBALQRfHgJHqy+TN6atZxeBRIWPmMbQjwz\ngQkKevw0GPJQckqOqJ6reZHtX1t3q4W0qFvB4aATZ2pW0h65T439KeGQJsF2SyRH\nJguyrrq69P+Htz5jb7n++piqi8MtlCmxeqsTZ0Q/An1qcVwNPi/5mO1G+iNBneOO\nqeMeHeJszesDAhUAhB//ijLvjwP7ytHCtytjvQzQbLsCgYEAsN1LTVxLr8THyGIN\nePKfDWzBYWW8eGhs7TIMuDv5m6ADYd5fNjGmFs/end4sV0YjQUlVo0scoFKWO5+9\nb91Pz1lQesT9B6UFpBg0dDisLj6Hcnle7FD9FOeV36OyRTuSZK506VjSvHofjRsA\n4+bh0MEZdKAs1gknIR0W2MX18E0DgYQAAoGAKoPMmd54NxqG7BjMwVpIeE3YhqOu\nC7r4GkCqf+Zs0ovllt/PucG6gyAULxcnIkhS60CK10GmrFzdTuRLqeUD8en6/gRu\nTiwGhjcq5in3o5KipoIhaQHkHE+k5TuhhqNqmZc9s602AkTm1ng4ggkganZiHM9A\nN4wXjPZE+VU7fqY=\n-----END PUBLIC KEY-----\n",
     "public_key_fingerprints": {
-        "blake2b": "96:3c:c2:e6:d5:c0:dd:0d:21:b2:04:1b:65:ab:fd:b9:8b:3d:a3:64:39:4f:1e:50:08:59:2f:79:c0:60:b6:8e:af:f6:b5:19:b6:f3:2d:76:4d:cc:4a:7e:df:a4:8a:74:17:0f:50:5d:5b:7e:b3:66:ab:3d:1b:5f:2f:8d:64:85",
-        "blake2s": "19:a7:ae:4c:27:de:38:e4:9e:84:27:d8:b6:fc:a3:57:89:86:09:2b:d4:6f:53:a3:67:71:87:fc:35:5a:6c:87",
-        "md5": "d0:c0:c6:32:d6:7c:eb:45:d1:d4:50:59:ec:00:d6:c3",
-        "sha1": "88:ca:a9:69:0b:cb:bc:9b:09:d0:ad:f3:3b:ba:26:93:f8:9d:50:e8",
-        "sha224": "5c:12:9f:87:83:6d:ed:c2:66:d2:b9:4d:58:fd:aa:02:2f:b1:16:c6:74:60:8d:62:f4:bf:ed:ee",
-        "sha256": "de:7d:dc:91:94:5d:7a:bb:d6:d5:18:ee:0a:7f:3f:ff:9b:c2:65:2a:34:b8:22:16:62:46:48:4d:45:56:52:a4",
-        "sha384": "f9:c1:e6:a2:06:98:06:d5:97:84:71:e0:1d:21:da:98:bf:7b:aa:c4:d9:96:e0:9e:ef:83:65:4d:54:9b:d4:6d:3c:20:75:af:bb:00:ab:74:61:40:cd:05:35:1f:65:4c",
-        "sha3_224": "26:85:88:09:a6:be:6e:bd:63:24:b4:90:fa:8b:0f:78:d8:ac:6e:0f:3b:8b:d6:8f:73:b9:47:94",
-        "sha3_256": "d7:70:54:59:e4:cc:6d:0e:e9:e1:8e:54:d3:69:48:4e:09:c5:fe:8a:7d:7c:df:f8:d4:39:01:8a:aa:50:01:02",
-        "sha3_384": "c4:10:cb:cb:ac:ea:76:5d:7a:a7:05:05:86:91:05:c9:a9:df:48:46:9a:75:d7:a1:c9:cd:30:27:25:fe:4c:84:24:58:b5:44:ca:54:fc:0b:2d:2a:21:34:1f:a2:00:31",
-        "sha3_512": "86:be:b6:4a:90:57:98:ba:d9:8c:4c:75:03:d5:f3:31:84:56:9c:db:36:c9:3d:4f:59:11:38:bf:df:04:bb:a0:e5:5f:e2:51:6e:5f:6f:60:2d:e6:7a:d1:2b:16:d8:16:84:8d:72:b9:a7:ec:8a:3b:1f:aa:2d:9b:da:20:9e:f8",
-        "sha512": "74:7c:27:ab:12:f2:8c:16:a3:6a:48:d0:c9:c8:e1:09:ab:e4:d6:a5:e4:03:71:b6:29:f9:04:96:ef:73:83:5d:c3:d3:1d:80:e1:34:89:58:4b:03:be:00:9f:30:38:80:fb:c2:7f:d1:c5:23:bb:31:0e:0d:f6:02:c5:25:5d:44",
-        "shake_128": "27:d3:81:4f:82:90:a5:69:77:3c:98:20:09:57:88:bd:2f:db:39:80:35:fd:b6:46:59:a4:02:30:59:f9:e5:f9",
-        "shake_256": "97:63:7a:53:b4:b5:da:8d:ea:67:aa:2c:a3:99:4a:d7:af:5c:05:d5:d6:90:fd:31:9d:a7:8c:6f:3c:c2:65:d6"
+        "blake2b": "1a:92:9d:2f:a4:d0:03:8b:ad:83:65:25:33:db:e5:8b:18:ab:0b:8f:cb:7a:4f:65:98:7e:4e:5d:ee:d9:7d:94:5b:10:d0:e8:b5:66:f5:7a:57:2f:ba:9a:64:73:44:40:76:af:b2:c3:25:ef:a4:0b:fc:71:66:a2:50:90:a8:4d",
+        "blake2s": "42:f5:08:68:cb:6c:1d:ae:49:88:8e:66:51:36:e7:04:03:d1:16:68:1c:01:ef:d7:b4:b7:c1:83:1e:74:5b:3f",
+        "md5": "29:38:c9:cc:b4:55:8f:d0:be:7f:ab:ed:1b:23:48:f7",
+        "sha1": "85:89:93:58:6a:7b:4a:e2:67:0a:60:2a:a2:f9:bb:fb:a2:87:12:a5",
+        "sha224": "b7:11:fb:07:a4:1d:44:57:d9:9c:cc:96:00:98:fe:82:c3:3a:ea:5b:c4:85:a7:b1:9f:15:da:74",
+        "sha256": "57:69:56:89:c8:22:e9:9b:84:43:8e:8d:db:36:6f:71:bd:51:7e:0a:38:96:bf:e6:02:bb:d9:d3:38:d3:6f:ed",
+        "sha384": "cc:c6:87:33:9d:d3:09:fb:83:43:61:8b:85:85:63:9e:bd:c2:ce:ca:9f:15:a8:59:e3:e5:a0:07:0a:84:21:05:56:44:ea:d7:ec:58:07:82:04:9c:34:74:57:40:25:cf",
+        "sha3_224": "df:74:0d:ee:05:52:29:95:58:cd:cc:2c:a7:0b:59:30:c6:25:33:7a:47:e7:d1:af:bc:82:6c:f4",
+        "sha3_256": "cb:b7:79:44:0a:9a:44:b6:ae:d5:d7:66:7e:15:84:e4:04:a0:8c:d2:d9:70:6c:af:56:d3:19:bb:91:c3:94:0c",
+        "sha3_384": "a6:73:6e:d9:36:08:e0:b6:41:dc:90:8f:c1:53:3d:c5:45:0e:da:30:30:71:4b:49:7a:dd:28:cb:13:63:22:7f:d6:5f:2e:48:88:d7:0c:c7:00:72:3b:25:78:98:75:de",
+        "sha3_512": "42:fc:86:1d:e8:f1:b4:e3:be:cf:f6:24:2d:8a:0b:f3:1c:2a:ec:ad:42:eb:f0:16:90:47:0f:53:23:58:d3:b9:50:e2:63:76:3c:04:e8:15:a7:48:09:16:05:f6:5f:3b:92:b8:b9:94:c1:b4:e2:c6:bd:65:08:cd:95:5b:ff:76",
+        "sha512": "0c:4c:0a:b3:4a:0d:9b:1c:fa:a1:a3:9a:53:64:9b:5e:20:dc:68:fc:50:77:49:5b:07:2d:a4:9a:a7:b3:2d:06:ef:d1:d0:c3:93:c6:3a:d4:7c:f0:e1:f0:0b:02:cb:4c:8c:ef:5b:96:8e:14:6d:24:00:66:17:a1:ca:50:71:a3",
+        "shake_128": "9b:68:ff:07:6a:7b:18:e3:80:03:11:bc:f4:51:a7:0c:7b:47:81:81:6d:9b:12:12:09:4d:e6:67:7c:ea:2d:77",
+        "shake_256": "84:91:ab:21:92:9c:3b:9b:b2:70:2f:c4:61:21:7d:57:48:f3:5e:8a:2b:ba:30:26:35:24:7c:8a:66:7f:86:a1"
     },
-    "type": "RSA"
+    "type": "DSA"
 }
```

**Comment by felixfontein** [comment]
For openssl_csr:
```.diff
TASK [(cryptography) Generate CSR with authority key identifier (change)] ************************************************************************************
--- before
+++ after
@@ -1,7 +1,7 @@
 {
     "authority_cert_issuer": null,
     "authority_cert_serial_number": null,
-    "authority_key_identifier": "00:11:22:33",
+    "authority_key_identifier": "44:55:66:77:88",
     "basic_constraints": null,
     "basic_constraints_critical": false,
     "can_parse_csr": false,
@@ -14,7 +14,7 @@
         },
         "2.5.29.35": {
             "critical": false,
-            "value": "MAaABAARIjM="
+            "value": "MAeABURVZneI"
         }
     },
     "key_usage": null,

changed: [localhost]
```

**Comment by felixfontein** [comment]
For x509_crl:
```.diff
TASK [Create CRL 2 (changed timestamps)] *********************************************************************************************************************
--- before
+++ after
@@ -12,8 +12,8 @@
             "Ansible"
         ]
     ],
-    "last_update": "20201130202531Z",
-    "next_update": "20201130202531Z",
+    "last_update": "20201130202532Z",
+    "next_update": "20201130202532Z",
     "revoked_certificates": [
         {
             "invalidity_date": null,
@@ -54,6 +54,16 @@
             "reason_critical": false,
             "revocation_date": "20201130202531Z",
             "serial_number": 1235
+        },
+        {
+            "invalidity_date": "20191012000000Z",
+            "invalidity_date_critical": false,
+            "issuer": null,
+            "issuer_critical": false,
+            "reason": "key_compromise",
+            "reason_critical": true,
+            "revocation_date": "20201130202532Z",
+            "serial_number": 185171575825867673923344777578182267298050875744
         }
     ]
 }

changed: [localhost]
```

**Comment by felixfontein** [comment]
For x509_certificate:
```.diff
TASK [(OwnCA, cryptography) Create authority key identifier (re-add)] ****************************************************************************************
--- before
+++ after
@@ -1,7 +1,7 @@
 {
     "authority_cert_issuer": null,
     "authority_cert_serial_number": null,
-    "authority_key_identifier": null,
+    "authority_key_identifier": "24:67:6d:71:6f:14:dc:76:72:4e:85:8e:10:0f:bc:ae:4c:4a:63:7c",
     "basic_constraints": null,
     "basic_constraints_critical": false,
     "can_parse_certificate": false,
@@ -16,23 +16,27 @@
         "2.5.29.17": {
             "critical": false,
             "value": "MBGCD3d3dy5leGFtcGxlLmNvbQ=="
+        },
+        "2.5.29.35": {
+            "critical": false,
+            "value": "MBaAFCRnbXFvFNx2ck6FjhAPvK5MSmN8"
         }
     },
     "fingerprints": {
-        "blake2b": "56:40:5b:f8:5a:b3:89:32:aa:43:1e:bb:48:d8:72:00:06:2f:b5:27:9f:22:30:37:46:90:6c:95:fe:8c:d3:48:09:98:d4:90:0c:c6:9b:59:d1:27:c0:18:4c:ed:d6:65:79:1c:b0:26:2d:ce:29:9d:6a:b8:90:9f:62:67:6b:d8",
-        "blake2s": "1a:04:ba:40:19:90:99:da:c9:fd:a2:95:f9:1f:34:f9:85:60:40:66:6e:2c:99:8c:50:75:67:22:91:9f:df:c2",
-        "md5": "2f:63:77:2b:4c:e8:95:e6:37:14:71:90:35:b2:6f:7c",
-        "sha1": "d0:30:93:63:5c:50:3c:cf:50:b9:a4:5f:66:38:28:61:eb:d2:c0:a0",
-        "sha224": "c0:c9:4b:d5:1b:87:55:f3:f1:04:3d:c3:81:97:b6:dd:37:1b:fa:83:4f:2e:b5:58:30:58:85:3a",
-        "sha256": "24:44:a1:6a:a4:d8:1c:c0:46:35:72:b0:6d:a2:2b:b6:04:bb:72:ce:f4:26:de:e8:8d:c7:ee:34:48:f9:76:d0",
-        "sha384": "77:3d:d0:eb:47:17:28:c8:0d:1e:40:27:6f:af:90:08:93:02:71:8c:6e:e7:20:79:d9:1d:df:6c:5d:66:3b:8e:db:ec:07:da:1d:2b:99:97:00:42:bf:ff:79:c6:b6:62",
-        "sha3_224": "44:1a:56:bc:20:f3:c3:0c:14:0f:aa:90:aa:1e:d1:0a:c8:95:64:47:78:92:2e:f5:a6:31:f7:e5",
-        "sha3_256": "68:c7:b6:57:98:4c:77:10:fb:a8:96:7c:a3:12:c0:fd:8b:c2:32:a3:bc:80:2d:14:09:cc:63:c3:4d:6d:0d:9e",
-        "sha3_384": "60:be:f1:51:70:7f:d2:0b:4c:6a:2a:c3:3a:7d:fd:4b:10:65:dc:53:c1:cf:78:aa:58:1f:f1:23:62:04:f5:67:50:21:a6:9e:31:d7:a7:46:00:d4:7f:43:2e:a5:c6:f2",
-        "sha3_512": "2d:c6:9b:66:22:7d:b8:89:5a:c5:0a:f9:20:a7:1b:72:fe:20:36:71:96:19:47:de:f1:8a:88:8d:7c:ed:04:ec:8a:98:96:f8:bf:8a:e0:b6:f3:23:b8:71:39:51:5d:10:f2:ef:be:cc:e3:2d:27:4e:5f:70:e6:85:d7:9a:10:58",
-        "sha512": "28:ec:d0:b0:d5:b4:23:8b:7f:2d:d9:b1:ea:c3:0b:6b:71:6c:4c:84:8f:13:3c:cd:a9:26:92:02:91:1e:d2:32:fc:dd:7e:a5:b0:07:8e:9c:f8:4f:70:23:bd:2f:fa:7c:c1:49:f8:f7:d8:11:84:ef:7b:85:fe:72:c4:23:47:b4",
-        "shake_128": "16:12:75:a7:18:7b:1c:fa:12:7e:45:3f:7e:3f:bf:55:61:a0:0f:53:fb:ea:c5:a7:5d:00:6e:10:e8:4b:f3:29",
-        "shake_256": "e2:59:1c:b1:a7:81:5e:7f:0e:86:54:1b:70:7d:64:48:fc:18:40:90:02:90:6d:bf:b3:e6:74:4e:2f:ae:4a:f8"
+        "blake2b": "ca:46:5b:b9:42:b5:1e:9f:ec:72:85:18:61:2b:b5:31:ff:a1:ac:f3:49:de:32:aa:1e:0e:36:51:53:af:9a:66:c9:d7:6d:78:aa:d4:9c:34:09:45:8a:af:15:81:2c:a4:24:7e:31:45:e7:5d:30:7c:ca:79:f1:bc:07:8e:ff:95",
+        "blake2s": "d0:33:45:a4:e8:f0:c7:e2:60:07:cf:76:06:9d:07:57:8c:65:49:13:cd:9f:9d:91:91:aa:9f:aa:2e:ff:50:4e",
+        "md5": "fd:1c:57:43:f8:37:a5:97:5d:86:fd:ce:ce:e8:fa:f8",
+        "sha1": "0a:ad:b1:ab:00:2b:2a:cf:a5:0b:46:62:ca:38:4c:8a:28:42:5e:87",
+        "sha224": "52:88:f1:a2:de:88:ac:18:ed:ee:4c:9a:ef:88:f6:f2:cf:b4:5c:e0:3c:68:2d:d8:86:72:63:fe",
+        "sha256": "81:f2:e5:f8:1e:d8:7c:ca:05:78:de:2f:97:42:e0:4e:35:5a:94:d7:4b:e9:f3:d4:e0:b3:d6:46:06:c8:0b:00",
+        "sha384": "36:9e:97:5e:93:ee:1a:4f:fc:50:87:33:cc:40:36:e5:ff:34:c9:19:69:3a:d7:f6:6d:0c:30:c3:ee:97:f8:15:29:c2:d0:32:ab:f1:e8:e5:7b:f3:34:05:0b:84:c0:86",
+        "sha3_224": "85:71:51:64:5c:8a:69:da:63:f5:80:03:85:57:93:9f:0c:46:3c:cb:48:95:59:5a:5c:64:6c:2d",
+        "sha3_256": "74:8e:43:66:e6:fb:b0:06:ca:68:71:c1:6d:a2:0b:4c:0e:b9:cf:73:56:ca:a2:10:43:b2:ba:96:5b:c1:66:56",
+        "sha3_384": "16:ba:bf:be:12:5c:57:17:12:05:22:25:cd:54:c2:64:e0:9b:45:35:dd:6d:f3:e0:34:42:31:8e:86:b1:91:5b:aa:51:6d:53:06:70:64:51:35:2c:4f:82:a0:41:f4:d4",
+        "sha3_512": "8e:cb:05:28:95:dd:ee:96:e5:28:9f:49:2a:85:be:90:93:a3:5b:fb:b8:9a:13:df:2d:74:40:2c:c6:63:40:48:8a:3a:3a:e4:19:7a:fc:87:37:36:8e:97:3d:24:1f:13:cb:f8:a6:fb:0a:9b:ce:78:c5:93:ff:1d:91:6c:d1:95",
+        "sha512": "bb:a6:35:c5:1b:49:e8:90:78:dd:93:d7:df:9c:2e:33:f4:da:09:d0:56:2e:cc:97:de:f4:5e:22:26:c4:93:36:5e:b6:c6:ce:9e:bd:75:e0:99:36:14:7c:1a:9f:29:1c:a6:7f:a9:bd:c7:9d:8c:d8:a8:2b:0a:e8:3d:f2:5d:f6",
+        "shake_128": "96:a6:b5:ca:57:3d:dc:37:28:42:2d:79:89:3f:0c:34:6f:3f:1f:90:8d:3f:7f:ea:1e:a0:f1:07:71:53:16:61",
+        "shake_256": "98:a6:aa:d2:1d:e8:4b:e9:c8:16:d2:f9:0e:74:55:94:30:22:48:01:b0:eb:0f:11:d4:5c:6e:ac:e8:67:42:4d"
     },
     "issuer": {
         "commonName": "Example CA"
@@ -45,8 +49,8 @@
     ],
     "key_usage": null,
     "key_usage_critical": false,
-    "not_after": "20301128203357Z",
-    "not_before": "20201130203357Z",
+    "not_after": "20301128203358Z",
+    "not_before": "20201130203358Z",
     "ocsp_must_staple": null,
     "ocsp_must_staple_critical": false,
     "ocsp_uri": null,
@@ -67,7 +71,7 @@
         "shake_128": "e6:5a:ac:08:e1:33:83:5e:5d:58:b9:a3:70:88:bf:60:87:1b:52:0c:bd:cf:a8:e9:a3:71:18:c2:ec:19:13:53",
         "shake_256": "eb:7c:f7:e3:f8:2f:77:a3:0f:2d:56:0c:89:9c:18:a4:26:86:89:7e:5f:3c:81:de:ff:ea:ff:4e:57:c6:64:aa"
     },
-    "serial_number": 195233888175763600194450600049968944137721032951,
+    "serial_number": 39194888486983301282758775737450223075141118303,
     "signature_algorithm": "sha256WithRSAEncryption",
     "subject": {
         "commonName": "www.example.com"

changed: [localhost]
```

**Comment by felixfontein** [comment]
ready_for_review

**Comment by felixfontein** [comment]
I added code which reduces the number of fingerprints used in diffs to one. This should reduce the excessive noise. For example:
```.diff
TASK [(cryptography) Generate privatekey_fmt_1 - auto format (no ignore)] ************************************************************************************
--- before
+++ after
@@ -3,12 +3,12 @@
     "key_is_consistent": true,
     "public_data": {
         "exponent": 65537,
-        "modulus": 770685813494779730665636395106638287162869989806318898065176400573021826902278223419571950267760807098665521308501255509203513133382033267470581356682000279112766565594879859007545093804443614452219014446454349878036243555601131461129910916084583478102722336612738153359937271243796567502801292364360902495688311020564395114874285331713293487882157108919193068539471556310636938179603902290586770886023742743910625326592483983548251697905291640667383278931559092951484453901342509956114335466887105022749313061859684524547330791578841820307061537264757715321689073182479093674019145099059871976050831313612618808229937965214406440156646006387756470575171443402705048081071693313392001839190465825015229355412830287588546302046972487821458430873447073175019734549656451656764179297018520997821849708088001203376794149565173147898657687061726035268837145227225518153627203601106463609596464293218997913752209428084169929196190638338245571718493130335639487197826708886156719173724906246601419742167125640964142411853252444189118365248532889936050106593969813205845772951146204800582850462045355611668489268997433079093752146863468007646623026141076756475956745194749815244492488574159768105031435426269510270991063801167080414865501601,
+        "modulus": 1012080207040108713178450916958556215901593760215493026515649809566227905775636537013789640868367704090499788172511898845765775724922663401556765710596861518084820531820730967313737929005192319560971817365325373739307934687814734388635005686397188114523401929892179913154444576027282249452605904630650230086390885932172541172037972612861409603108523007265651329815343458478945685264141653786494079647919694852310633554407259400160016660756138477209133430438678130219275941155914959215921374554898971350490691390260583757909721433955819721751761535796091461956698353577346542642674310591743295057538529295851090593259059224333121404201464622793199292771446309038404825089866705801976444409283777520550722316513978271099028369194988198436542825408254470403054044859963960268462854363926500045974506414433487556589066328182814978339937895034226939986793542552291413072953482540049859174074230339174001286465756430649476348235867553793437563084215651451959839111740206177629141307481029246101409160579471537500240428581604770596761124678282801090632901648105883217996569364055751031200649620369570156312785322250368674062091189829861940913377371088818414816617056379987583935785753801551406782485721924354972763180124216928416132730650863,
         "size": 4096
     },
-    "public_key": "-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAvOj50WDuQgl4fIn/5ONy\neh9nM70FBsZA5UwGZBYPPC3ks0+IocB62COqsNVbRPF8tKJzT1nPtxlg0egIcfpV\n+TQdd5JePMPFRB2ShW3a0zOk3TGMKxjJ1hQn/fYe9+/kBz1I6IchxrZSB/Hn9cNq\no84k+HCTaNqgUImjtR4zPxkTUkwT4p0p7sqMvClY4p0c1VrklPsRXqg4o8iSjkcA\n1dKEWaHIrOfcxAChiDqgoe+Yl7gmrLlTOyiYQjSxdIZaZbF1Q2eoBBzD5k5r3Y0n\n5R0wraYjG5vCbzxJXipg6nl+xMykXb6LUZb8k7AsKtrnLc0OG7XxIdFSKjqVYTqL\nwXWyDvU/urmU4OL8hfb7r22hRsA3dhNENEKMS4PTqGnRZxv7n+hrpIHmyvV/dIs3\njwk2iS5ND0I8iLTDOTM44tMtTkrhWqdY6aMz9eFb/69Cp7O+MfVzQNHv5Ealk9Ch\nSAqbS6FKdvfKTP02rfzR6YEFnZos5g5ukR8nHu59AYqprdCwregb7aN6CwkD3exx\nfq0yJjW4pUWEkDWItkCrREU3vlKfn8wmHjyq3SKPAsc1EnbYh6EZzrTDE3HpJEse\nhvYdctCpxMXJXilpQiLOXBmEJxYasmyokteZQLHPfmXI1UMwpgtEGZ9YTz/WRGas\n/VvceNVS5TLY1dcY/YE/raECAwEAAQ==\n-----END PUBLIC KEY-----\n",
+    "public_key": "-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA+BScuLHqQvjHQerj1/4k\n1g4WGzV1hqcP97CHaU+IVYMUeYC8DHO5bIDIgzNK9AElhh/XEs9mBfE1bb+UrI30\nWjQ9dl87eVu1ywE99vNG4DMOVuQTOrJjm6gKNNd5w4GEgqnl/EitLeo2qRP213Iy\nLms64Wdv5O+/w0b8B7pRO7pbUW6WWr+voyofNlJjoaOEaiNd761+taz3gLOK3Jku\nauqgI2wwqZvXNQqqKvQdU/ldO3p+lbSJB5PejLvegKw8wEYAR5maEWK2Ey5z21VN\nCR2x9PBNswBbSaM4BI64AS7ZZy89JQM3UQevsloPcg4obQ8GJLjnJxnAN9mAxHG6\nZ7CaYWgGftZP62duCpphhg9KWl3pTtDrpb+9MXAECeoJqe+MhXRXWABXl1vTBTuq\ndykNVwDhznrt82m/BoHsamS3HsPxTEbUvelDpHSCeUR2M4hAqGVl8hfhRhlG9vLR\n4ProYI+7LQtYzWLi1JEj7Y4JHo7WvWJ8xhHUC7RsU0FDQi7zOCDgbXxABj6RVEqj\nprM3TQUthLW9wbwZpFM1ALAXivJhBL1hiOZGQxrd0PXts+wzYWS08sCX+B3fuk9N\nW7xWuf+PSxWIB5oztzT46xc2OnDSTiQmFdZUKFyg4H5MryZbrdDogk1ZGcRFQ+fa\nrtHQgvczBEkKu/SvFoEHAO8CAwEAAQ==\n-----END PUBLIC KEY-----\n",
     "public_key_fingerprints": {
-        "sha256": "b1:cf:5e:79:31:ae:96:09:1a:d3:11:b7:76:96:71:94:22:03:f2:1d:30:8b:0b:2a:f0:62:d6:55:ad:d8:25:b8"
+        "sha256": "06:56:cc:90:35:a4:cf:c8:41:f1:29:54:b5:77:12:0d:c1:cc:ce:4f:fa:c2:dd:8e:73:11:6a:04:e6:e3:28:ff"
     },
     "type": "RSA"
 }
```

**Comment by felixfontein** [comment]
One thing that this PR does not do is provide diff in two cases:
1. When used with check_mode. To provide a diff in that case, the object in question would have to be regenerated for that, which would be an unnecessary computation. We could try to detect the combination diff + check mode and only recompute them in that case (but not return / store the result), but that seems a bit excessive. And as opposed to other modules, we also cannot easily 'fake' how the diff would look like without writing a lot new (error-prone) code.
2. When changing key format in `openssl_privatekey`, or when the key format is not PEM. That's mainly caused by the private key info functionality only supporting PEM keys. That should be fixed/improved in a follow-up PR.

**Review by kbatdorf37 — APPROVED** [review]

**Comment by felixfontein** [comment]
I've created PRs for the first four commits: #203, #204, #205, #206. Once these are reviewed and merged, I'll rebase this one.

**Comment by felixfontein** [comment]
ready_for_review

File: plugins/modules/x509_crl.py [hunk 1]
```diff
@@ -550,6 +554,19 @@ def __init__(self, module):
...
+            data = None
+
+        self.diff_after = self.diff_before = self._get_info(data)
+
+    def _get_info(self, data):
+        if data is None:
+            return dict()
+        try:
+            result = get_crl_info(self.module, data)
+            result['can_parse_crl'] = False
```
Comment (Ajpantuso) [inline plugins/modules/x509_crl.py:566]:
```suggestion
            result['can_parse_crl'] = True
```

**Review by Ajpantuso — CHANGES_REQUESTED** [review]
Only comments are for the `can_parse_xxx` result which seems inconsistent among each module as is.

File: plugins/module_utils/crypto/module_backends/certificate.py [hunk 2]
```diff
@@ -95,6 +99,19 @@ def __init__(self, module, backend):
...
 
+        self.diff_before = self._get_info(None)
+        self.diff_after = self._get_info(None)
+
+    def _get_info(self, data):
+        if data is None:
+            return dict()
+        try:
+            result = get_certificate_info(self.module, self.backend, data, prefer_one_fingerprint=True)
+            result['can_parse_certificate'] = False
```
[the remark this answers is not in the document]
Comment (felixfontein) [reply plugins/module_utils/crypto/module_backends/certificate.py:110]: No. I have no idea what I was thinking here :)

**Review by Ajpantuso — APPROVED** [review]
LGTM, thanks!

**Comment by felixfontein** [comment]
@kbatdorf37 @Ajpantuso thanks for reviewing this!

**Pull request merged**